### PR TITLE
API: libpacemaker: Add function for listing resource agents.

### PR DESCRIPTION
**Pull request by clumens**
This is the same functionality as "crm_resource --list-agents", but in a pair of public and private functions.  crm_resource has then been modified to just call the private function instead of doing this on its own.

Ref T44

### CODE REVIEW

**Comment by clumens** [comment]
This is the first, very easy, step towards converting crm_resource into a front end for a bunch of library functions.

A couple other things I thought of while working on this:

* include/pacemaker.h is going to get big and should probably be reorganized into sections and functions alphabetized within those sections.
* I don't know that any of our public functions have callers at the moment, which means I don't know that any of the public functions have really been tested.  That's definitely worth looking into.

**Comment by clumens** [comment]
Looks like this is the only user:

> tools/crm_simulate.c:562:    pcmk_free_injections(options.injections);

So yeah, I guess I'm a little concerned about how well these functions currently work.

File: lib/pacemaker/pcmk_agents.c [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ */
+int
+pcmk__list_agents(pcmk__output_t *out, const char *agent_spec)
+{
+    int rc = pcmk_rc_ok;
+    char *provider = strchr(agent_spec, ':');
+    lrmd_t *lrmd_conn = NULL;
+    lrmd_list_t *list = NULL;
+
+    CRM_ASSERT(out != NULL);
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_agents.c:33]: might as well assert agent_spec too (and do the strchr() after)

File: lib/pacemaker/pcmk_agents.c [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    rc = lrmd__new(&lrmd_conn, NULL, NULL, 0);
+    if (rc != pcmk_rc_ok) {
+        goto error;
+    }
+
+    if (provider) {
+        *provider++ = 0;
+    }
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_agents.c:42]: agent_spec should probably not be const (or we could strdup() it)

File: tools/crm_resource.c [hunk 3]
```diff
@@ -1860,7 +1819,7 @@ main(int argc, char **argv)
             break;
 
         case cmd_list_agents:
-            rc = list_agents(out, options.agent_spec);
+            rc = pcmk__list_agents(out, options.agent_spec);
```
Comment (kgaillot) [inline tools/crm_resource.c:1822]: Will we end up showing errors twice, once in the function and once in `pcmk__output_and_clear_error()`? See `done:`

**Review by kgaillot — COMMENTED** [review]
good start :)

File: tools/crm_resource.c (hunk 3 above)
Comment (clumens) [reply tools/crm_resource.c:1822]: Yes and no:
```
[root@rhel9-cluster-1 ~]# crm_resource --list-agents xyz
No agents found for standard 'xyz'
crm_resource: Error performing operation: Error occurred
```

**Comment by clumens** [comment]
Added another commit on top to take care of the other --list-* options.

File: tools/crm_resource.c (hunk 3 above)
Comment (kgaillot) [reply tools/crm_resource.c:1749]: I suppose that's fine then. Though I would expect "no such object" in this particular example ...

Comment (clumens) [reply tools/crm_resource.c:1749]: We could certainly revisit the error codes here, too.  "Error occurred" is really unhelpful, and I suppose that's coming from just returning pcmk_rc_error.  That's what list_agents was doing and I just copied it over.

Comment (kgaillot) [reply tools/crm_resource.c:1749]: In the off chance someone's script is checking for exit status 1, let's hold that for a compatibility break, but I think it's a good idea

File: lib/pacemaker/pcmk_agents.c [hunk 4]
```diff
@@ -83,3 +134,110 @@ pcmk_list_agents(xmlNodePtr *xml, char *agent_spec)
...
+
+    if (rc > 0) {
+        rc = out->message(out, "standards-list", list);
+    } else {
+        rc = pcmk_rc_error;
+    }
+
+error:
+    if (rc != pcmk_rc_ok) {
+       out->err(out, _("No %s found"), "OCF providers");
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_agents.c:218]: "standards"

**Review by kgaillot — COMMENTED** [review]
looking good

File: lib/pacemaker/pcmk_agents.c (hunk 4 above)
Comment (clumens) [reply lib/pacemaker/pcmk_agents.c]: Fixed.

**Review by kgaillot — APPROVED** [review]

**Pull request merged**